jax/experimental/host_callback.py

```diff
@@ -0,0 +1,305 @@
+# limitations under the License.
+"""Implementation of an experimental primitive for printing, including
+   from transformed and compiled code.
+
+See documentation for `id_print` below.
+For usage example, see tests/host_callback_test.py.
+
+Implementation plan:
+  * Write the API for the `id_print` primitive, using data-dependence as
+    explained in `id_print` documentation (DONE).
+  * Implement the transformations. DONE (except pmap)
+  * Implement the JIT for CPU using CustomCall in C++. DONE (except unit tests
+    do not run in OSS; also missing float16 and bfloat16).
+  * Implement the JIT for GPU using also CustomCall in C++. DONE.
+  * Explore how to pipe the printed data back to the Colab cell,
+    when running in Colab. ?
+  * Explore implementation using outfeed, hoping that it works for all
+    platforms, and can pipe data more easily. STARTED.
+  * Explore feeding the data back to the Python program (the `id_tap`
+    primitive). ?
+  * Explore a simpler API that uses Python program-order, instead of
+    data dependency-order. Need to add support to JAX for stateful primitives.
+"""
+
+from functools import partial
+import io
+
+from jax import lax, core
+from jax.lib import pytree, xla_bridge
+from jax.interpreters import ad, xla, batching
+from jax.interpreters import partial_eval as pe
+from jaxlib import xla_client
+from jaxlib import xla_extension
+
+import msgpack  # type: ignore
+import numpy as onp
+import os
+from typing import Any, Dict, List, Sequence, Tuple
+
+if os.getenv("JAX_ENABLE_JIT_PRINT", "false") != "false":
+  from jaxlib import host_callback_cpu_py
+  for _name, _value in host_callback_cpu_py.customcall_registrations().items():
+    xla_client.register_custom_call_target(_name, _value, platform="cpu")
+
+  try:
+    from jaxlib import host_callback_cuda_py  # type: ignore[import-error]
+    for _name, _value in host_callback_cuda_py.customcall_registrations().items():
+      xla_client.register_custom_call_target(_name, _value, platform="gpu")
+  except ImportError:
+    pass  # We may have built without CUDA support
+
+# TODO(necula): fix mypy errors if I define the type aliases below
+XlaOp = Any  # xla_extension.XlaOp
+XlaShape = Any # xla_client.Shape 
+XlaComputationBuilder = Any  # xla_bridge._JaxComputationBuilder
+
+id_print_p = core.Primitive("id_print")
+id_print_p.multiple_results = True
+
+
+def id_print(*args, **kwargs):
+  """Behaves like the identify function for positional arguments, but prints all
+     arguments on the host, even from transformed or compiled code.
+
+     The return value is a tuple with the value of `args` or the value of the
+     keyword parameter `result` if present. If there is a single positional
+     argument, it returns just that argument without packing it in a tuple.
+
+     The positional arguments must be JAX values. The keyword arguments are
+     serialized to a string and printed along with the positional arguments.
```

Comment:
Would it be feasible to support strings inline in `*args` instead? The rule might be that strings are dropped from `args` in the default return value, e.g,. so you can write something like `x = id_print('x =', x)` or `x, y = id_print('x =', x, 'y =', y)`.

Using the keyword argument for extra strings is a little more awkward to write, and it isn't 100% clear that only strings values are OK. I would sort of expect `id_print(what=x)` to print `what=` followed by the value of `x`.

Reply:
These are all good suggestions, I will play with them, I suspect that much of this can be added to the API.